shell/platform/embedder/embedder_surface_gl_impeller.cc

```diff
@@ -0,0 +1,198 @@
+
+  FML_LOG(ERROR) << "Using the Impeller rendering backend (GLES/ANGLE).";
+  valid_ = true;
+}
+
+EmbedderSurfaceGLImpeller::~EmbedderSurfaceGLImpeller() = default;
+
+// |EmbedderSurface|
+bool EmbedderSurfaceGLImpeller::IsValid() const {
+  return valid_;
+}
+
+// |GPUSurfaceGLDelegate|
+std::unique_ptr<GLContextResult>
+EmbedderSurfaceGLImpeller::GLContextMakeCurrent() {
+  worker_->SetReactionsAllowedOnCurrentThread(true);
+  return std::make_unique<GLContextDefaultResult>(
+      gl_dispatch_table_.gl_make_current_callback());
+}
+
+// |GPUSurfaceGLDelegate|
+bool EmbedderSurfaceGLImpeller::GLContextClearCurrent() {
+  worker_->SetReactionsAllowedOnCurrentThread(false);
+  return gl_dispatch_table_.gl_clear_current_callback();
+}
+
+// |GPUSurfaceGLDelegate|
+bool EmbedderSurfaceGLImpeller::GLContextPresent(
+    const GLPresentInfo& present_info) {
+  // Pass the present information to the embedder present callback.
+  return gl_dispatch_table_.gl_present_callback(present_info);
+}
+
+// |GPUSurfaceGLDelegate|
+GLFBOInfo EmbedderSurfaceGLImpeller::GLContextFBO(
+    GLFrameInfo frame_info) const {
+  // Get the FBO ID using the gl_fbo_callback and then get exiting damage by
+  // passing that ID to the gl_populate_existing_damage.
+  return gl_dispatch_table_.gl_populate_existing_damage(
+      gl_dispatch_table_.gl_fbo_callback(frame_info));
+}
+
+// |GPUSurfaceGLDelegate|
+bool EmbedderSurfaceGLImpeller::GLContextFBOResetAfterPresent() const {
+  return fbo_reset_after_present_;
+}
+
+// |GPUSurfaceGLDelegate|
+SkMatrix EmbedderSurfaceGLImpeller::GLContextSurfaceTransformation() const {
+  auto callback = gl_dispatch_table_.gl_surface_transformation_callback;
+  if (!callback) {
+    SkMatrix matrix;
+    matrix.setIdentity();
+    return matrix;
+  }
+  return callback();
+}
+
+// |GPUSurfaceGLDelegate|
+EmbedderSurfaceGL::GLProcResolver EmbedderSurfaceGLImpeller::GetGLProcResolver()
+    const {
+  return gl_dispatch_table_.gl_proc_resolver;
+}
+
+// |GPUSurfaceGLDelegate|
+SurfaceFrame::FramebufferInfo
+EmbedderSurfaceGLImpeller::GLContextFramebufferInfo() const {
+  // Enable partial repaint by default on the embedders.
+  auto info = SurfaceFrame::FramebufferInfo{};
+  info.supports_readback = true;
+  info.supports_partial_repaint =
+      gl_dispatch_table_.gl_populate_existing_damage != nullptr;
+  return info;
+}
+
+// |EmbedderSurface|
+std::unique_ptr<Surface> EmbedderSurfaceGLImpeller::CreateGPUSurface() {
+  const bool render_to_surface = !external_view_embedder_;
+  return std::make_unique<GPUSurfaceGLImpeller>(
+      this,              // GPU surface GL delegate
+      impeller_context_  // render to surface
+  );
+}
+
+// |EmbedderSurface|
+std::shared_ptr<impeller::Context>
+EmbedderSurfaceGLImpeller::CreateImpellerContext() const {
+  return impeller_context_;
+}
+
+// |EmbedderSurface|
+sk_sp<GrDirectContext> EmbedderSurfaceGLImpeller::CreateResourceContext()
+    const {
+  sk_sp<GrDirectContext> resource_context;
```

Comment:
This seems wrong but I'm not sure how else to make it work. 🤷‍♂️

Reply:
I don't think you need this.

You should just see if you can make the resource context current, and if you can mark the worker as allowed to react on this thread. And then always return nullptr.

Returning a valid context from here should not be required and we should not be creating a GrContext when using Impeller.

Reply:
Ahh OK, that works and makes much more sense. I briefly read through the code and thought this still needed to be constructed

Reply:
The resource context is a performance improving optional thing. Impeller handles things differently by just using worker threads and the reactor mechanism - Skia OTOH expects you to manage a context for the IO thread.

Returning nullptr here should be tolerated anywhere. We may eventually want some better way to signal what this method is actually doing but right now it's a little hard to think about the best way to clean up all the duplicated branches where we've baked Skia concepts into things (like methods that return nullptr for GrContexts in various spots, or nullptr for impeller::Contexts in others).